### DISCUSSION TITLE
Target edge runtime and server components conditionally

### DISCUSSION THREAD

**Discussion by benmerckx** [vercel/next.js#39094] [Ideas]
### Describe the feature you'd like to request

Api routes configured with `{runtime: 'experimental-edge'}` target `"browser"` exports during bundling.
Libraries that use [conditional exports](https://nodejs.org/api/packages.html#conditional-exports) might want to export different code for server and client. Currently it seems there's no way to target the edge api routes specifically. For users this can result in build errors such as `Dynamic Code Evaluation (e. g. 'eval', 'new Function', 'WebAssembly.compile') not allowed in Edge Runtime` for unwanted browser specific code.

Cloudflare workers via wrangler have the option of targetting the [`"worker"` export](https://github.com/cloudflare/wrangler2/issues/84).

### Describe the solution you'd like

Provide a target which enables libraries to conditionally provide exports for the edge runtime. Potentially re-use the `"worker"` condition.

````js
{
  "exports": {
    // Target edge runtime
    "worker": "./server.js",
    // Actual browsers
    "browser": "./client.js",
    // Everything else
    "default": "./server.js"
  }
}
````

### Describe alternatives you've considered

Don't see other options.

**Comment by ItsWendell** [comment]
React (Server) and Emotion actually does support the same thing, it would be great to add this!

**Comment by benmerckx** [comment]
Not sure if anyone monitors these discussions or if this is where issues come to die. I just ran into a situation where it would be beneficial to have an option to target server components conditionally as well. In Next 13 the compiler fails hard when something like useState is imported from React regardless whether that import is actually used at runtime or not.

**Comment by ValentinH** [comment]
I just tried it and it seems that the Edge Runtime already supports the `worker` exports key. 🙂

**Comment by benmerckx** [comment]
Thanks for letting me know!

**Comment by Swimburger** [comment]
Is there a condition specific to Vercel Edge Runtime, not just the generic `worker`?

**Comment by benmerckx** [comment]
There is the "edge-light" export now. But make sure it comes before other exports like "browser".
